### PR TITLE
[model, ops] refactor: add NPU support and OpSlot guard for Qwen3/VL/MoE, Qwen3.5/MoE

### PR DESCRIPTION
**Pull request by yanghw116**
### What does this PR do?

Qwen3/VL/MoE, Qwen3.5/MoE模型适配昇腾NPU卡

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> Validation results (training curves, eval metrics) for changes not covered by CI.

### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

参考GPU patch方案，增加NPU patch代码。
主要更改点：
1、没有NPU patch代码的模型会新建NPU patch代码，已有的则修改。
2、每个模型的大部分结构都保持跟GPU一致，需要自定义算子的部分则相应修改。
3、自定义算子使用OpSlot注册。

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: veomni/models/transformers/qwen3_5/qwen3_5_npu_patch_gen_config.py
Comment (yicheng-gong) [inline veomni/models/transformers/qwen3_5/qwen3_5_npu_patch_gen_config.py]: using opslot same as gpu

Comment (yicheng-gong) [inline veomni/models/transformers/qwen3_5/qwen3_5_npu_patch_gen_config.py]: change to import torch_npu here

Comment (yanghw116) [reply veomni/models/transformers/qwen3_5/qwen3_5_npu_patch_gen_config.py]: done. please refer to VeOmni/veomni/ops/kernels/gated_delta_rule/npu_rms_norm_gated.py

Comment (yanghw116) [reply veomni/models/transformers/qwen3_5/qwen3_5_npu_patch_gen_config.py]: done

File: veomni/ops/kernels/rotary/npu_rotary_pos_emb.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    q_embed = torch_npu.npu_rotary_mul(q_rot, cos, sin)
+    k_embed = torch_npu.npu_rotary_mul(k_rot, cos, sin)
+
+    # Concatenate back to full shape
+    q_embed = torch.cat([q_embed, q_pass], dim=-1)
+    k_embed = torch.cat([k_embed, k_pass], dim=-1)
+    return q_embed, k_embed
+
+
+def partial_apply_rotary_pos_emb_vision(q: torch.Tensor, k: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor):
```
Comment (yicheng-gong) [inline veomni/ops/kernels/rotary/npu_rotary_pos_emb.py]: rename due to here is not partial rope

File: veomni/ops/kernels/rotary/npu_rotary_pos_emb.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+def full_apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+    cos = cos.unsqueeze(unsqueeze_dim)
+    sin = sin.unsqueeze(unsqueeze_dim)
+    q_embed = torch_npu.npu_rotary_mul(q, cos, sin)
+    k_embed = torch_npu.npu_rotary_mul(k, cos, sin)
+    return q_embed.to(q.dtype), k_embed.to(k.dtype)
+
+
+def full_apply_rotary_pos_emb_vision(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
Comment (yicheng-gong) [inline veomni/ops/kernels/rotary/npu_rotary_pos_emb.py]: Can this be merged into the function defined on line 23? It seems both reshape BSH from 3D to 4D for the operator

File: veomni/ops/kernels/rotary/npu_rotary_pos_emb.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import torch
```
Comment (yicheng-gong) [inline veomni/ops/kernels/rotary/npu_rotary_pos_emb.py]: An npu.py file with similar functionality already exists, so we don’t need a new file.

File: veomni/ops/kernels/rotary/npu_rotary_pos_emb.py (hunk 1 above)
Comment (yanghw116) [reply veomni/ops/kernels/rotary/npu_rotary_pos_emb.py]: done

File: veomni/ops/kernels/rotary/npu_rotary_pos_emb.py (hunk 2 above)
Comment (yanghw116) [reply veomni/ops/kernels/rotary/npu_rotary_pos_emb.py]: done

File: veomni/ops/kernels/rotary/npu_rotary_pos_emb.py (hunk 3 above)
Comment (yanghw116) [reply veomni/ops/kernels/rotary/npu_rotary_pos_emb.py]: done

File: veomni/models/transformers/qwen3/generated/patched_modeling_qwen3_npu.py [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+@use_kernel_func_from_hub("rotary_pos_emb")
+def apply_rotary_pos_emb(q, k, cos, sin, unsqueeze_dim=1):
```
Comment (yicheng-gong) [inline veomni/models/transformers/qwen3/generated/patched_modeling_qwen3_npu.py]: It seems opslot is not taking effect here

Comment (yanghw116) [reply veomni/models/transformers/qwen3/generated/patched_modeling_qwen3_npu.py]: decorator is removed.

File: veomni/models/transformers/qwen3_vl_moe/qwen3_vl_moe_npu_patch_gen_config.py [hunk 5]
```diff
@@ -161,66 +166,13 @@
...
-    return q_embed, k_embed
-    # --- Patch.1 ---
-
-
-# ================================================================
-# Patch: Qwen3VLMoeTextRMSNorm.forward -> NPU fused npu_rms_norm
-# 1. swap the full-fp32 variance path for `torch_npu.npu_rms_norm`
-#    which stays in the weight dtype and is significantly faster on NPU
-# ================================================================
-@config.override_method(
```
Comment (FoolPlayer) [inline veomni/models/transformers/qwen3_vl_moe/qwen3_vl_moe_npu_patch_gen_config.py:224]: Why remove this

File: veomni/models/transformers/qwen3_5_moe/qwen3_5_moe_npu_patch_gen_config.py [hunk 6]
```diff
@@ -156,12 +156,6 @@
     description="Use fused rmsnorm to impl zero-centered rmsnorm (1+weight centered formulation)",
 )
 
-config.override_method(
```
Comment (FoolPlayer) [inline veomni/models/transformers/qwen3_5_moe/qwen3_5_moe_npu_patch_gen_config.py:163]: Same here and other models.

File: veomni/ops/kernels/rotary/npu.py [hunk 7]
```diff
@@ -37,21 +38,31 @@ def apply_rotary_pos_emb_npu(q, k, cos, sin, position_ids=None, unsqueeze_dim=1)
 
 
 def apply_rotary_pos_emb_vision_npu(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
-    orig_q_shape = q.shape
-    orig_k_shape = k.shape
-    orig_q_dtype = q.dtype
-    orig_k_dtype = k.dtype
-    q_4d = q.unsqueeze(0).float().contiguous()
```
Comment (FoolPlayer) [inline veomni/ops/kernels/rotary/npu.py:44]: Why remove transfering q/k to float?

**Review by FoolPlayer — CHANGES_REQUESTED** [review]

**Comment by FoolPlayer** [comment]
And:
1. Stale rms_norm_gated_implementation help text. Still claims "Qwen3.5 has no NPU backend today" even though this PR adds one, and doesn't list "npu" as a valid value.

2. Missing Apache 2.0 header on the new npu_rms_norm_gated.py.

3. Misleading description= strings like "OpSlot guard for NPU fused RMSNorm" on dispatchers that don't actually have an NPU kernel behind them

File: veomni/models/transformers/qwen3_vl_moe/qwen3_vl_moe_npu_patch_gen_config.py (hunk 5 above)
Comment (yanghw116) [reply veomni/models/transformers/qwen3_vl_moe/qwen3_vl_moe_npu_patch_gen_config.py:224]: not removed. just import from qwen3_vl_gpu_patch_gen_config and reuse the patch code.

File: veomni/models/transformers/qwen3_5_moe/qwen3_5_moe_npu_patch_gen_config.py (hunk 6 above)
Comment (yanghw116) [reply veomni/models/transformers/qwen3_5_moe/qwen3_5_moe_npu_patch_gen_config.py:163]: rmsnorn gated is moved to veomni/ops/kernels/gated_delta_rule/npu_rms_norm_gated.py, and registered in opslot. you can choose which kernel you like in yaml

File: veomni/ops/kernels/rotary/npu.py (hunk 7 above)
Comment (yanghw116) [reply veomni/ops/kernels/rotary/npu.py:44]: in npu_rotary_mul, q/k will automatically be transferred to float. so we don't need to do it outside. but cos and sin are different. so they should be transferred in advance.

**Comment by yanghw116** [comment]
> And:
> 
> 1. Stale rms_norm_gated_implementation help text. Still claims "Qwen3.5 has no NPU backend today" even though this PR adds one, and doesn't list "npu" as a valid value.
> 2. Missing Apache 2.0 header on the new npu_rms_norm_gated.py.
> 3. Misleading description= strings like "OpSlot guard for NPU fused RMSNorm" on dispatchers that don't actually have an NPU kernel behind them

1. fixed.
2. done.
3. npu kernel is moved to veomni/ops/kernels/rms_norm/npu.py and registered in init.py

File: veomni/ops/kernels/rotary/npu.py (hunk 7 above)
Comment (FoolPlayer) [reply veomni/ops/kernels/rotary/npu.py:44]: Get. But at the end, do we need to cast the type back?

Comment (yanghw116) [reply veomni/ops/kernels/rotary/npu.py:44]: don'n have to. cos and sin are only used within this function.

**Review by FoolPlayer — APPROVED** [review]
